File: go/test_helpers.go
package parser

import (
	"encoding/json"
	"fmt"
	"slices"
	"strings"
	"sync"
)

var (
	successorsMu sync.Mutex
	successors   = make(map[string]string)
)

// -- helper functions

func ptrString(s string) *string {
	return &s
}

func buildExpectedKV(k string, v interface{}, expected map[string]interface{}) {
	b, _ := json.Marshal(v)
	expected[k] = bytesToValue(b)
}

func feedChunksToParser(cp *ChunkParser, chunks [][]byte) *int {
	total := len(chunks)
	matched := 0
	overflown := 0
	var result *int

	for i, chunk := range chunks {
		fmt.Printf("Processing chunk %d/%d bytes %d\n", i+1, len(chunks), len(chunk))
		cp.ProcessChunk(chunk, i == total-1)
		if len(cp.DoneFields) > matched {
			fmt.Printf("Matched so far: %v\n", cp.DoneFields)
			matched = len(cp.DoneFields)
		}
		if len(cp.OverflowedFields) > overflown {
			fmt.Printf("Overflow so far: %v\n", cp.OverflowedFields)
			overflown = len(cp.OverflowedFields)
		}
		if cp.isAllFound() {
			idx := i + 1
			result = &idx
			return result
		}
	}
	return nil
}

func buildExpectedWithPos(pathMap map[string][2]interface{}, chunks [][]byte) map[string][2]interface{} {
	expected := make(map[string][2]interface{})
	for jsonPath, output := range pathMap {
		fieldsVec := strings.Split(jsonPath, ".")
		rem := []byte{}
		for i, chunk := range chunks {
			data := append([]byte{}, rem...)
			data = append(data, chunk...)
			jsonValue := extractJSONValue(data, fieldsVec)
			if jsonValue != nil {
				outputKey := output[0].(*string)
				if outputKey == nil {
					outputKey = &jsonPath
				}
				expected[*outputKey] = [2]interface{}{i + 1, jsonValue}
				break
			}
			rem = data
		}
	}
	fmt.Println("expected:", len(expected))
	return expected
}

func buildExpected(pathMap map[string][2]interface{}, allExpected map[string]interface{}) map[string]interface{} {
	expected := make(map[string]interface{})
	for jsonPath, output := range pathMap {
		if jsonValue, ok := allExpected[jsonPath]; ok {
			outputKey := output[0].(*string)
			if outputKey == nil {
				outputKey = &jsonPath
			}
			expected[*outputKey] = jsonValue
		}
	}
	return expected
}

func getExpectedFieldNames(pathMap map[string][2]interface{}) ([]string, map[string]string) {
	var fields [][]string
	var allNames []string
	successorMap := make(map[string]string)

	for k := range pathMap {
		fields = append(fields, strings.Split(k, "."))
	}

	for _, f := range fields {
		for _, name := range f {
			if !slices.Contains(allNames, name) {
				allNames = append(allNames, name)
			}
		}
	}

	successorsMu.Lock()
	for _, name := range allNames {
		if val, ok := successors[name]; ok {
			successorMap[name] = val
		}
	}
	successorsMu.Unlock()

	return allNames, successorMap
}

func v(ch rune, valueLen int) string {
	return fmt.Sprintf(`"%s"`, rep(ch, valueLen))
}

func textValue(ch rune, valueLen int, quote rune) string {
	q := string(quote)
	return fmt.Sprintf("%s%s%s", q, rep(ch, valueLen), q)
}

func buildText(fieldCount, valueLen int, quote rune, kvsep string, sep rune, expected *map[string]interface{}) string {
	var flatFields strings.Builder
	q := string(quote)

	for i := 0; i < fieldCount; i++ {
		ch := rune('a' + (i % 26))
		key := fmt.Sprintf("field_%d", i)
		value := rep(ch, valueLen)
		flatFields.WriteString(fmt.Sprintf(`%s%s%s%s%s%s%s`, q, key, q, kvsep, q, value, q))
		flatFields.WriteRune(sep)
		if expected != nil {
			(*expected)[key] = bytesToValue([]byte(fmt.Sprintf(`"%s"`, value)))
		}
	}
	return flatFields.String()
}

func buildFlatFields(fieldCount, valueLen int, expected map[string]interface{}) string {
	return buildText(fieldCount, valueLen, '"', ": ", ',', &expected)
}

func strArray(ch rune, itemCount, valueLen int) string {
	items := make([]string, itemCount)
	for i := 0; i < itemCount; i++ {
		items[i] = v(ch, valueLen)
	}
	return fmt.Sprintf("[ %s ]", strings.Join(items, ","))
}

func objArray(kch, vch rune, itemCount, valueLen int) string {
	items := make([]string, itemCount)
	for i := 0; i < itemCount; i++ {
		items[i] = fmt.Sprintf(`{"name":%s,"value":%s}`, v(kch, valueLen), v(vch, valueLen))
	}
	return fmt.Sprintf("[ %s ]", strings.Join(items, ","))
}

func setLargerJSONSuccessors() {
	successorsMu.Lock()
	defer successorsMu.Unlock()

	m := map[string]string{
		"field_0": "field_1", "field_1": "field_2", "field_2": "field_3",
		"field_3": "field_4", "field_4": "field_5", "field_5": "field_6",
		"field_6": "field_7", "field_7": "field_8", "field_8": "field_9",
		"field_9": "field_10", "field_10": "field_11", "field_11": "field_12",
		"field_12": "field_13", "field_13": "field_14", "field_14": "field_15",
		"field_15": "field_16", "field_16": "field_17", "field_17": "field_18",
		"field_18": "field_19", "field_19": "field_20", "field_20": "field_21",
		"field_21": "field_22", "field_22": "field_23", "field_23": "field_24",
		"field_24": "field_25", "field_25": "field_26", "field_26": "field_27",
		"field_27": "field_28", "field_28": "field_29", "field_29": "field_30",
		"field_30": "field_31", "metadata": "author", "author": "version",
		"version": "stats", "stats": "views", "views": "details",
		"details": "regions", "regions": "locale", "locale": "name",
		"name": "tags", "tags": "items", "items": "config", "config": "key",
		"key": "values", "values": "signature", "signature": "timestamp",
	}
	successors = m
}

func buildLargeJSON(fieldCount, valueLen int, verbose bool, allExpected map[string]interface{}) []byte {
	expected := make(map[string]interface{})
	flatFields := buildFlatFields(fieldCount, valueLen, expected)
	setLargerJSONSuccessors()

	smallJSON := buildSmallJSON(fieldCount, valueLen, allExpected)

	// Remove trailing '}' and leading '{' from smallJSON to embed it
	smallJSONStr := string(smallJSON)
	if len(smallJSONStr) >= 2 {
		smallJSONStr = smallJSONStr[1 : len(smallJSONStr)-1]
	}

	json := fmt.Sprintf(
		"{%s%s}",
		flatFields,
		smallJSONStr,
	)

	jsonBytes := []byte(json)

	// Merge flat fields into allExpected
	for k, v := range expected {
		allExpected[k] = v
	}

	if verbose {
		fmt.Println("=== JSON structure ===")
		printJSONStructure(jsonBytes)
		fmt.Println("=== end structure ===")
	}
	return jsonBytes
}

func buildSmallJSON(fieldCount, valueLen int, allExpected map[string]interface{}) []byte {
	successorsMu.Lock()
	successors = map[string]string{
		"metadata": "author", "author": "version", "version": "stats",
		"stats": "views", "views": "details", "details": "regions",
		"regions": "locale", "locale": "name", "name": "tags",
		"tags": "items", "items": "config", "config": "key",
		"key": "values", "values": "signature", "signature": "timestamp",
	}
	successorsMu.Unlock()

	author := textValue('A', valueLen, ' ')
	version := textValue('B', valueLen, ' ')
	views := objArray('C', 'D', fieldCount, valueLen)
	regions := strArray('E', fieldCount, valueLen)
	locale := textValue('F', valueLen, ' ')
	name := textValue('G', valueLen, ' ')
	tags := strArray('H', fieldCount, valueLen)
	items := objArray('I', 'J', fieldCount, valueLen)
	key := textValue('K', valueLen, ' ')
	value1 := textValue('L', valueLen, ' ')
	value2 := textValue('M', valueLen, ' ')
	signature := strArray('N', fieldCount, valueLen)
	timestamp := 123456789

	// toStr marshals a Go string to a JSON string then unmarshals back, matching
	// Rust's to_json_value(&string) which serializes the String type.
	toStr := func(s string) interface{} {
		b, _ := json.Marshal(s)
		return bytesToValue(b)
	}

	// Format string matches Rust's build_small_json exactly:
	//   - "name" is inside "metadata" (after "stats" closes)
	//   - "", "items", "config", "signature", "timestamp" are all at root level
	jsonStr := fmt.Sprintf(
		`{`+
			` "metadata" :  {`+
			`   ""  :  "%s"  ,`+
			`   "author"  :  "%s"  ,`+
			`"version":"%s",`+
			`"stats":{`+
			`"views":%s,`+
			` "details" :   {  `+
			`  "regions"   :   %s,`+
			`  "locale"   :  "%s"  `+
			`}}`+  // close details, close stats
			`,`+   // comma after stats
			`"name": "%s"`+ // name INSIDE metadata (after stats)
			`},`+  // close metadata, comma
			`"":%s,`+
			`"items":%s,`+
			`"config":{`+
			`"key": "%s",`+
			`"values":{`+
			`"value1": "%s",`+
			`"value2": "%s"`+
			`}}`+ // close values, close config
			`,`+  // comma
			`"signature":%s,`+
			`"timestamp":%d`+
			`}`, // close root
		author, author, version, views, regions, locale, name,
		tags, items, key, value1, value2, signature, timestamp,
	)

	jsonBytes := []byte(jsonStr)

	allExpected["metadata."] = toStr(author)
	allExpected["metadata.author"] = toStr(author)
	allExpected["metadata.version"] = toStr(version)
	allExpected["metadata.stats.views"] = bytesToValue([]byte(views))
	allExpected["metadata.stats.details.regions"] = bytesToValue([]byte(regions))
	allExpected["metadata.stats.details.locale"] = toStr(locale)
	allExpected["metadata.name"] = toStr(name)
	allExpected[""] = bytesToValue([]byte(tags))
	allExpected["items"] = bytesToValue([]byte(items))
	allExpected["config.key"] = toStr(key)
	allExpected["config.values.value1"] = toStr(value1)
	allExpected["config.values.value2"] = toStr(value2)
	allExpected["signature"] = bytesToValue([]byte(signature))
	allExpected["timestamp"] = bytesToValue([]byte(fmt.Sprintf("%d", timestamp)))

	fmt.Println("=== JSON structure ===")
	printJSONStructure(jsonBytes)
	fmt.Println("=== end structure ===")
	return jsonBytes
}

func buildInvalidJSON(fieldCount, valueLen int, allExpected map[string]interface{}) []byte {
	json := buildSmallJSON(fieldCount, valueLen, allExpected)
	json = append(json, []byte(`-------`)...)
	json = append(json, []byte(`+++++++`)...)
	return json
}

func buildTextInput(fieldCount, valueLen int, expected *map[string]interface{}) []byte {
	text := []byte(buildText(fieldCount/3, valueLen, ' ', " ", ' ', expected))
	text = append(text, []byte(`-------`)...)
	text = append(text, []byte(buildText(fieldCount/3, valueLen, ' ', " ", ' ', expected))...)
	text = append(text, []byte(`+++++++`)...)
	text = append(text, []byte(buildText(fieldCount/3, valueLen, ' ', " ", ' ', expected))...)
	return text
}

func randomChunks(bytes []byte, minSize, maxSize int, seed uint64, splitRandomKeys bool, keys []string) [][]byte {
	var chunks [][]byte
	pos := 0
	rng := seed
	lastSplit := false

	for pos < len(bytes) {
		r := maxSize - minSize
		size := minSize + int(lcgNext(&rng)%uint64(maxInt(r, 1)))
		end := minInt(pos+size, len(bytes))
		chunk := bytes[pos:end]

		if splitRandomKeys {
			didSplit := false
			if !lastSplit {
				lastSplit = true
				content := string(chunk)
				for _, key := range keys {
					if idx := strings.Index(content, key); idx >= 0 {
						mid := idx + len(key)/2 + 1
						if mid < len(chunk) {
							chunks = append(chunks, chunk[:mid])
							chunks = append(chunks, chunk[mid:])
							didSplit = true
							break
						}
					}
				}
			} else {
				lastSplit = false
			}
			if !didSplit {
				chunks = append(chunks, chunk)
			}
		} else {
			chunks = append(chunks, chunk)
		}
		pos = end
	}
	return chunks
}

func bytesToValue(b []byte) interface{} {
	var v interface{}
	json.Unmarshal(b, &v)
	return v
}

func extractJSONValue(bytes []byte, path []string) interface{} {
	parser := NewJSONEventGenerator()
	if len(path) == 0 {
		return nil
	}

	cursor := 0
	var pendingKey *string
	matchedDepth := 0
	skippedDepth := 0
	collectingDepth := 0
	collectStart := 0

	for cursor < len(bytes) {
		sliceToParse := bytes[cursor:]
		w := parser.NextEvent(sliceToParse, true)
		cursor += w.ConsumedBytes

		if w.Event == nil {
			if w.ConsumedBytes == 0 {
				break
			}
			continue
		}

		if w.Err != nil {
			break
		}

		ev := w.Event
		if ev.Kind == EventEof {
			break
		}

		if ev.Kind == EventObjectKey {
			if skippedDepth == 0 && collectingDepth == 0 {
				pendingKey = &ev.String
			}
		}

		if ev.Kind == EventStartObject {
			if collectingDepth > 0 {
				collectingDepth++
			} else {
				key := ""
				if pendingKey != nil {
					key = *pendingKey
					pendingKey = nil
				}

				if skippedDepth > 0 {
					skippedDepth++
				} else if key == "" {
					// Root — transparent
				} else if matchedDepth < len(path) && key == path[matchedDepth] {
					if matchedDepth == len(path)-1 {
						collectingDepth = 1
						collectStart = cursor - 1
					} else {
						matchedDepth++
					}
				} else {
					skippedDepth++
				}
			}
		}

		if ev.Kind == EventEndObject {
			if collectingDepth > 0 {
				collectingDepth--
				if collectingDepth == 0 {
					return bytesToValue(bytes[collectStart:cursor])
				}
			} else {
				if skippedDepth > 0 {
					skippedDepth--
				} else if matchedDepth > 0 {
					matchedDepth--
				}
				pendingKey = nil
			}
		}

		if ev.Kind == EventStartArray {
			if collectingDepth > 0 {
				collectingDepth++
			} else {
				if skippedDepth > 0 {
					skippedDepth++
				} else if matchedDepth < len(path) &&
					pendingKey != nil && *pendingKey == path[matchedDepth] &&
					matchedDepth == len(path)-1 {
					collectingDepth = 1
					collectStart = cursor - 1
				} else {
					skippedDepth++
				}
				pendingKey = nil
			}
		}

		if ev.Kind == EventEndArray {
			if collectingDepth > 0 {
				collectingDepth--
				if collectingDepth == 0 {
					return bytesToValue(bytes[collectStart:cursor])
				}
			} else if skippedDepth > 0 {
				skippedDepth--
			}
		}

		if ev.Kind == EventString {
			if collectingDepth == 0 && skippedDepth == 0 &&
				matchedDepth == len(path)-1 &&
				pendingKey != nil && *pendingKey == path[matchedDepth] {
				b, _ := json.Marshal(ev.String)
				return bytesToValue(b)
			}
			if collectingDepth == 0 {
				pendingKey = nil
			}
		}

		if ev.Kind == EventNumber {
			if collectingDepth == 0 && skippedDepth == 0 &&
				matchedDepth == len(path)-1 &&
				pendingKey != nil && *pendingKey == path[matchedDepth] {
				return bytesToValue([]byte(ev.String))
			}
			if collectingDepth == 0 {
				pendingKey = nil
			}
		}

		if ev.Kind == EventBoolean {
			if collectingDepth == 0 && skippedDepth == 0 &&
				matchedDepth == len(path)-1 &&
				pendingKey != nil && *pendingKey == path[matchedDepth] {
				val := "false"
				if ev.Boolean {
					val = "true"
				}
				return bytesToValue([]byte(val))
			}
			if collectingDepth == 0 {
				pendingKey = nil
			}
		}
	}
	return nil
}

func lcgNext(state *uint64) uint64 {
	*state = *state*6364136223846793005 + 1442695040888963407
	return *state
}

func rep(ch rune, length int) string {
	return strings.Repeat(string(ch), length)
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: lua/test_helpers.lua
-- test_helpers.lua
-- Mirrors rust/tests/test_helpers.rs
-- Utility functions for building test JSON and exercising the chunk parser.

local parser    = require("parser")
local cp_module = require("chunk_parser")

local M = {}

-- ── LCG random number generator ────────────────────────────────────────────
-- 32-bit Knuth LCG — works in LuaJIT (doubles) and Lua 5.3+ (integers).
-- Result is always in [0, 2^32-1], so no sign-masking needed.
function M.lcg_next(state)
    return (state * 1664525 + 1013904223) % 0x100000000
end

-- ── String helpers ──────────────────────────────────────────────────────────

function M.rep(ch, len)
    return string.rep(ch, len)
end

-- Build a value string: characters repeated, optionally quoted.
function M.text_value(ch, value_len, quote)
    local inner = M.rep(ch, value_len)
    return quote .. inner .. quote
end

-- ── JSON building helpers ───────────────────────────────────────────────────

function M.str_array(ch, item_count, value_len)
    local items = {}
    for _ = 1, item_count do
        items[#items + 1] = M.text_value(ch, value_len, '"')
    end
    return "[ " .. table.concat(items, ",") .. " ]"
end

function M.obj_array(kch, vch, item_count, value_len)
    local items = {}
    for _ = 1, item_count do
        items[#items + 1] = string.format('{"name":%s,"value":%s}',
            M.text_value(kch, value_len, '"'),
            M.text_value(vch, value_len, '"'))
    end
    return "[ " .. table.concat(items, ",") .. " ]"
end

-- Build flat scalar fields: "field_0": "aaa...", "field_1": "bbb...", ...
-- Populates all_expected[field_name] = decoded_value.
function M.build_flat_fields(field_count, value_len, all_expected)
    local parts = {}
    for i = 0, field_count - 1 do
        local ch = string.char(string.byte('a') + (i % 26))
        local key = string.format("field_%d", i)
        local value = M.rep(ch, value_len)
        parts[#parts + 1] = string.format('"%s": "%s"', key, value)
        if all_expected then
            -- Stored as a plain Lua string (the JSON string value)
            all_expected[key] = value
        end
    end
    return table.concat(parts, ",") .. ","
end

-- Build the small JSON structure (matches Rust's build_small_json exactly).
-- Populates all_expected with paths → values.
function M.build_small_json(field_count, value_len, all_expected)
    local author    = M.text_value('A', value_len, ' ')
    local version   = M.text_value('B', value_len, ' ')
    local views     = M.obj_array('C', 'D', field_count, value_len)
    local regions   = M.str_array('E', field_count, value_len)
    local locale    = M.text_value('F', value_len, ' ')
    local name      = M.text_value('G', value_len, ' ')
    local tags      = M.str_array('H', field_count, value_len)
    local items     = M.obj_array('I', 'J', field_count, value_len)
    local key       = M.text_value('K', value_len, ' ')
    local value1    = M.text_value('L', value_len, ' ')
    local value2    = M.text_value('M', value_len, ' ')
    local signature = M.str_array('N', field_count, value_len)
    local timestamp = 123456789

    local json = string.format(
        '{'..
        ' "metadata" :  {'..
        '   ""  :  "%s"  ,'..
        '   "author"  :  "%s"  ,'..
        '"version":"%s",'..
        '"stats":{'..
        '"views":%s,'..
        ' "details" :   {  '..
        '  "regions"   :   %s,'..
        '  "locale"   :  "%s"  '..
        '}},'..
        '"name": "%s"'..
        '},'..
        '"":%s,'..
        '"items":%s,'..
        '"config":{'..
        '"key": "%s",'..
        '"values":{'..
        '"value1": "%s",'..
        '"value2": "%s"'..
        '}},'..
        '"signature":%s,'..
        '"timestamp":%d'..
        '}',
        author, author, version,
        views, regions, locale, name,
        tags, items, key, value1, value2,
        signature, timestamp
    )

    if all_expected then
        -- Scalar strings: stored as raw leaf values (what the parser returns)
        -- The parser strips JSON quotes, so we store the unquoted content.
        -- text_value uses space as quote, so " FFF " → parser returns " FFF ".
        all_expected["metadata."]                     = author
        all_expected["metadata.author"]               = author
        all_expected["metadata.version"]              = version
        -- Container values: stored as raw JSON strings (parser decodes to table)
        all_expected["metadata.stats.views"]          = views
        all_expected["metadata.stats.details.regions"]= regions
        all_expected["metadata.stats.details.locale"] = locale
        all_expected["metadata.name"]                 = name
        all_expected[""]                              = tags
        all_expected["items"]                         = items
        all_expected["config.key"]                    = key
        all_expected["config.values.value1"]          = value1
        all_expected["config.values.value2"]          = value2
        all_expected["signature"]                     = signature
        -- Numbers stored as Lua numbers (parser returns tonumber())
        all_expected["timestamp"]                     = timestamp
    end

    return json
end

-- Build the large JSON (flat fields + small JSON structure).
function M.build_large_json(field_count, value_len, all_expected)
    local flat = M.build_flat_fields(field_count, value_len, all_expected)
    local small = M.build_small_json(field_count, value_len, all_expected)
    -- Remove outer braces from small JSON to embed it
    small = string.sub(small, 2, #small - 1)  -- strip leading '{' and trailing '}'
    return "{" .. flat .. small .. "}"
end

-- Build invalid JSON (small JSON + garbage appended).
function M.build_invalid_json(field_count, value_len, all_expected)
    local json = M.build_small_json(field_count, value_len, all_expected)
    return json .. "-------" .. "+++++++"
end

-- Build plain text input (not valid JSON).
local function build_text_section(field_count, value_len)
    local parts = {}
    for i = 0, field_count - 1 do
        local ch = string.char(string.byte('a') + (i % 26))
        local key = string.format("field_%d", i)
        local value = M.rep(ch, value_len)
        parts[#parts + 1] = string.format(" %s  %s ", key, value)
    end
    return table.concat(parts, " ")
end

function M.build_text_input(field_count, value_len)
    local third = math.floor(field_count / 3)
    return build_text_section(third, value_len)
        .. "-------"
        .. build_text_section(third, value_len)
        .. "+++++++"
        .. build_text_section(third, value_len)
end

-- ── Chunk splitting ─────────────────────────────────────────────────────────

-- Split `bytes` (a string) into random-sized chunks [min, max).
-- If split_random_keys is true, occasionally splits in the middle of a key.
-- seed is the LCG seed (integer).
function M.random_chunks(bytes, min_size, max_size, seed, split_random_keys, keys)
    local chunks = {}
    local pos = 1  -- 1-based into bytes string
    local rng = seed
    local last_split = false
    local len = #bytes

    while pos <= len do
        local range = max_size - min_size
        rng = M.lcg_next(rng)
        -- rng is always in [0, 2^32-1] so modulo is always non-negative
        local size = min_size + (rng % math.max(range, 1))
        local end_pos = math.min(pos + size - 1, len)
        local chunk = string.sub(bytes, pos, end_pos)

        local did_split = false
        if split_random_keys and keys and #keys > 0 and not last_split then
            last_split = true
            for _, key in ipairs(keys) do
                if key ~= "" then
                    local ki = string.find(chunk, key, 1, true)
                    if ki then
                        local mid = ki + math.floor(#key / 2)
                        if mid < #chunk then
                            chunks[#chunks + 1] = string.sub(chunk, 1, mid)
                            chunks[#chunks + 1] = string.sub(chunk, mid + 1)
                            did_split = true
                            break
                        end
                    end
                end
            end
        else
            if split_random_keys then
                last_split = false
            end
        end

        if not did_split then
            chunks[#chunks + 1] = chunk
        end

        pos = end_pos + 1
    end

    return chunks
end

-- ── Expected value extraction ───────────────────────────────────────────────

-- Extract the JSON value at a dot-path from a JSON string using the event generator.
-- path_parts is a list of string keys, e.g. {"metadata", "stats", "locale"}.
-- Returns a Lua value (via decode_json_value), or nil.
function M.extract_json_value(json_str, path_parts)
    if #path_parts == 0 then return nil end

    local gen = parser.JSONEventGenerator.new()
    local cursor = 0
    local len = #json_str

    local pending_key    = nil
    local matched_depth  = 0
    local skipped_depth  = 0
    local collecting_depth = 0
    local collect_start  = 0  -- 0-based byte position of start of collected region

    while cursor <= len do
        local slice = string.sub(json_str, cursor + 1)
        local w = gen:next_event(slice, true)
        cursor = cursor + w.consumed_bytes

        if w.event == nil then
            if w.consumed_bytes == 0 then break end
        end

        if w.error and not w.event then break end

        local ev = w.event
        if ev == nil then break end
        local ek = ev.kind

        if ek == parser.EVENT_EOF then break end

        if ek == parser.EVENT_OBJECT_KEY then
            if skipped_depth == 0 and collecting_depth == 0 then
                pending_key = ev.value
            end

        elseif ek == parser.EVENT_START_OBJECT then
            if collecting_depth > 0 then
                collecting_depth = collecting_depth + 1
            else
                local k = pending_key or ""
                pending_key = nil
                if skipped_depth > 0 then
                    skipped_depth = skipped_depth + 1
                elseif k == "" then
                    -- root object — transparent
                elseif matched_depth < #path_parts and k == path_parts[matched_depth + 1] then
                    if matched_depth == #path_parts - 1 then
                        collecting_depth = 1
                        collect_start = cursor - 1  -- points at '{'
                    else
                        matched_depth = matched_depth + 1
                    end
                else
                    skipped_depth = skipped_depth + 1
                end
            end

        elseif ek == parser.EVENT_END_OBJECT then
            if collecting_depth > 0 then
                collecting_depth = collecting_depth - 1
                if collecting_depth == 0 then
                    local raw = string.sub(json_str, collect_start + 1, cursor)
                    return cp_module.decode_json_value(raw)
                end
            else
                if skipped_depth > 0 then
                    skipped_depth = skipped_depth - 1
                elseif matched_depth > 0 then
                    matched_depth = matched_depth - 1
                end
                pending_key = nil
            end

        elseif ek == parser.EVENT_START_ARRAY then
            if collecting_depth > 0 then
                collecting_depth = collecting_depth + 1
            else
                local k = pending_key
                pending_key = nil
                if skipped_depth > 0 then
                    skipped_depth = skipped_depth + 1
                elseif matched_depth < #path_parts
                    and k ~= nil and k == path_parts[matched_depth + 1]
                    and matched_depth == #path_parts - 1
                then
                    collecting_depth = 1
                    collect_start = cursor - 1  -- points at '['
                else
                    skipped_depth = skipped_depth + 1
                end
            end

        elseif ek == parser.EVENT_END_ARRAY then
            if collecting_depth > 0 then
                collecting_depth = collecting_depth - 1
                if collecting_depth == 0 then
                    local raw = string.sub(json_str, collect_start + 1, cursor)
                    return cp_module.decode_json_value(raw)
                end
            elseif skipped_depth > 0 then
                skipped_depth = skipped_depth - 1
            end

        elseif ek == parser.EVENT_STRING then
            if collecting_depth == 0 and skipped_depth == 0
                and matched_depth == #path_parts - 1
                and pending_key ~= nil and pending_key == path_parts[matched_depth + 1]
            then
                return ev.value  -- plain Lua string
            end
            if collecting_depth == 0 then pending_key = nil end

        elseif ek == parser.EVENT_NUMBER then
            if collecting_depth == 0 and skipped_depth == 0
                and matched_depth == #path_parts - 1
                and pending_key ~= nil and pending_key == path_parts[matched_depth + 1]
            then
                return tonumber(ev.value)
            end
            if collecting_depth == 0 then pending_key = nil end

        elseif ek == parser.EVENT_BOOLEAN then
            if collecting_depth == 0 and skipped_depth == 0
                and matched_depth == #path_parts - 1
                and pending_key ~= nil and pending_key == path_parts[matched_depth + 1]
            then
                return ev.value
            end
            if collecting_depth == 0 then pending_key = nil end

        else
            if collecting_depth == 0 then pending_key = nil end
        end
    end

    return nil
end

-- Build expected results from path_map + all_expected.
-- path_map: { [path] = { output_key_or_nil, max_size } }
-- all_expected: { [path] = raw_value }
function M.build_expected(path_map, all_expected)
    local expected = {}
    for path, opts in pairs(path_map) do
        local v = all_expected[path]
        if v ~= nil then
            local out_key = opts[1] or path
            expected[out_key] = v
        end
    end
    return expected
end

-- Build expected with the chunk index where the value first became readable.
function M.build_expected_with_pos(path_map, chunks)
    local expected = {}
    for json_path, opts in pairs(path_map) do
        local parts = {}
        for part in (json_path .. "."):gmatch("([^.]*)%.") do
            parts[#parts + 1] = part
        end
        local accumulated = ""
        for i, chunk in ipairs(chunks) do
            accumulated = accumulated .. chunk
            local v = M.extract_json_value(accumulated, parts)
            if v ~= nil then
                local out_key = opts[1] or json_path
                expected[out_key] = { i, v }
                break
            end
        end
    end
    return expected
end

-- Get all field name components for chunk-splitting heuristic.
function M.get_expected_field_names(path_map)
    local seen = {}
    local names = {}
    for path in pairs(path_map) do
        for part in (path .. "."):gmatch("([^.]*)%.") do
            if part ~= "" and not seen[part] then
                seen[part] = true
                names[#names + 1] = part
            end
        end
    end
    return names
end

-- Feed chunks to the parser; return the 1-based index of the chunk where
-- all targets were found, or nil.
function M.feed_chunks_to_parser(cp, chunks)
    local total = #chunks
    for i, chunk in ipairs(chunks) do
        cp:process_chunk(chunk, i == total)
        if cp:is_all_found() then
            return i
        end
    end
    return nil
end

-- ── Deep equality ────────────────────────────────────────────────────────────

-- Compare two Lua values that may have come from JSON decoding.
-- Strings from all_expected are raw JSON-string contents (no quotes),
-- but extract_json_value also returns the raw string content — so comparison
-- is direct string equality for strings.
function M.values_equal(a, b)
    if type(a) ~= type(b) then
        -- Allow number/string comparison for numeric values
        if type(a) == "number" and type(b) == "string" then
            return tostring(a) == b or a == tonumber(b)
        elseif type(a) == "string" and type(b) == "number" then
            return a == tostring(b) or tonumber(a) == b
        end
        return false
    end
    if type(a) == "table" then
        if a == cp_module.NULL and b == cp_module.NULL then return true end
        for k, v in pairs(a) do
            if k ~= "is_array" then
                if not M.values_equal(v, b[k]) then return false end
            end
        end
        for k in pairs(b) do
            if k ~= "is_array" and a[k] == nil then return false end
        end
        return true
    end
    return a == b
end

return M
